### ISSUE TITLE
Periodic sustained massive slowdown in inception training

### ISSUE THREAD

**Issue by liuyipei** [tensorflow/models#170]
I am running the inception network on [2x TitanX, ubuntu 14.04, cuda 7.5, tensorflow 0.8.0, cudnn 5.0.5, built from source]. Inception starts start off well, but usually after a day in, I start getting these massive slow-downs. During these slowdowns, CPU and GPU are mostly idle. There is no thermal issue, IO is idle, and there is a single thread that runs at 100% (with intermittent activity on other cores every few seconds). 

These slowdowns are intermittent and can last for hours at a time, effectively reducing training throughput by more than half.

How would I go about diagnosing this issue? I think this is evidence that there is some single-threaded part in the pipeline that somehow starts off fine, but starts accumulating more workload overtime during training. What is very shocking is that these slowdowns last for hours, but somehow the speed will pick back up again at another time.

The nvidia-smi output during a slowdown is shown below.

```
+------------------------------------------------------+                       
| NVIDIA-SMI 352.93     Driver Version: 352.93         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:02:00.0      On |                  N/A |
| 88%   50C    P2    96W / 250W |  11675MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:81:00.0      On |                  N/A |
| 88%   51C    P2    85W / 250W |  11763MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      1506    G   /usr/bin/X                                     239MiB |
|    0      2545    G   compiz                                         210MiB |
|    0     21340    C   /usr/bin/python                              11193MiB |
|    1      1506    G   /usr/bin/X                                      21MiB |
|    1     21340    C   /usr/bin/python                              11710MiB |
+-----------------------------------------------------------------------------+
```
Here is a representative workload as represented by `top` during one of the slowdowns. I know there is always a single 100% thread from `htop` (not shown here). (i.e. I know that the 114% is not an aggregate of many medium sized threads -- there is always one approximately 100%).
```
top - 16:23:28 up 12 days,  1:40,  7 users,  load average: 8.46, 7.22, 7.12
Tasks: 416 total,   1 running, 414 sleeping,   0 stopped,   1 zombie
%Cpu(s):  0.6 us,  4.5 sy,  0.0 ni, 94.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  65842856 total, 65089280 used,   753576 free,   181136 buffers
KiB Swap:  9765884 total,   251544 used,  9514340 free. 56480068 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND     
21340 yiliu     20   0  0.105t 8.629g 2.262g S 114.2 13.7  14938:09 python      
28312 root      20   0   58888  17132   7280 S   4.6  0.0   0:30.44 iotop       
 1506 root      20   0  351808  74044  54972 S   3.3  0.1 121:17.50 Xorg        
28151 yiliu     20   0   26824   4652   2996 S   2.7  0.0   0:16.61 htop        
27071 yiliu     20   0  649820  33672  22020 S   2.3  0.1   0:08.48 gnome-term+ 
 2545 yiliu     20   0 1896360 180048  82460 S   1.3  0.3  32:59.72 compiz      
  120 root      rt   0       0      0      0 S   0.3  0.0   0:02.52 watchdog/16 
28352 yiliu     20   0 1293116 303012  84636 S   0.3  0.5   1:42.46 firefox     
    1 root      20   0   33852   3956   2576 S   0.0  0.0   0:03.54 init        
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.26 kthreadd    
    3 root      20   0       0      0      0 S   0.0  0.0   0:14.93 ksoftirqd/0 
```

Here is a typical output during the start of a slowdown: note how it drops from 49 examples to 3-5.
```
2016-06-01 13:10:35.564939: global_step_scalar 140290, loss = 5.76 (49.0 examples/sec; 1.306 sec/batch) 
2016-06-01 13:11:14.245180: global_step_scalar 140300, loss = 4.81 (6.3 examples/sec; 10.144 sec/batch) 
2016-06-01 13:13:55.796055: global_step_scalar 140310, loss = 4.82 (3.0 examples/sec; 21.507 sec/batch) 
2016-06-01 13:16:39.183686: global_step_scalar 140320, loss = 5.25 (2.5 examples/sec; 25.218 sec/batch) 
2016-06-01 13:20:24.135745: global_step_scalar 140330, loss = 4.35 (3.7 examples/sec; 17.442 sec/batch) 
2016-06-01 13:23:32.838517: global_step_scalar 140340, loss = 5.32 (2.4 examples/sec; 26.164 sec/batch) 
2016-06-01 13:26:28.802064: global_step_scalar 140350, loss = 5.43 (3.3 examples/sec; 19.354 sec/batch) 
2016-06-01 13:29:35.901208: global_step_scalar 140360, loss = 4.75 (2.9 examples/sec; 22.128 sec/batch) 
2016-06-01 13:32:22.282453: global_step_scalar 140370, loss = 5.02 (3.1 examples/sec; 20.379 sec/batch) 
2016-06-01 13:35:13.818376: global_step_scalar 140380, loss = 4.95 (5.3 examples/sec; 12.053 sec/batch) 
2016-06-01 13:37:53.258600: global_step_scalar 140390, loss = 5.21 (5.7 examples/sec; 11.153 sec/batch) 
2016-06-01 13:40:59.375350: global_step_scalar 140400, loss = 4.88 (2.0 examples/sec; 32.358 sec/batch) 
2016-06-01 13:44:49.049245: global_step_scalar 140410, loss = 5.15 (3.6 examples/sec; 17.882 sec/batch) 
2016-06-01 13:47:15.656218: global_step_scalar 140420, loss = 4.38 (4.9 examples/sec; 13.164 sec/batch) 
2016-06-01 13:49:46.658372: global_step_scalar 140430, loss = 5.02 (3.6 examples/sec; 17.671 sec/batch) 
2016-06-01 13:51:58.691818: global_step_scalar 140440, loss = 5.59 (3.6 examples/sec; 17.927 sec/batch) 
```

**Comment by andydavis1** [comment]
Can you follow the instructions referenced in this github issue to get a timeline that captures the slow down?
 https://github.com/tensorflow/tensorflow/issues/3009

**Comment by FangxiangFeng** [comment]
I have the same problem.

**Comment by heibaidaolx123** [comment]
@andydavis1 Hi, I tried adding timeline code to inception_distributed_train.py, but the output json was empty like:
{
    "traceEvents": [
        {
            "ph": "M",
            "args": {
                "name": "Allocators"
            },
            "pid": 0,
            "name": "process_name"
        }
    ]
}.
 And adding timeline code to inception_train.py, the output json is OK and viewable on Chrome. So does the distributed code support timeline? Or did I miss something?

**Comment by hetaoaoao** [comment]
I have the exact same issue.
The following picture is from tensorboard event tab. If the batch join queue is full, train performance drop down. It looks like the consumer doesn't have full power to process the queue, but gpu usage was low at the same time, I don't get it.
Another finding out, increase image preprocess thread count would make this issue happen more offen, on my side.
![image](https://cloud.githubusercontent.com/assets/16871995/16893636/f961fee8-4b70-11e6-86bd-a2c3ecf3e911.png)

**Comment by hetaoaoao** [comment]
@andydavis1 
here's the timeline. 
Issue #3099 mentioned that it's slow, but on my case, it was quick on most cases, and get slow down sometime later. I also post a timeline when it's on "normal" situation.

Any clues?  it's really frustrating
![image](https://cloud.githubusercontent.com/assets/16871995/16895227/0ba53bb6-4ba2-11e6-8d0c-f0e078599720.png)

![image](https://cloud.githubusercontent.com/assets/16871995/16895257/2f455852-4ba3-11e6-9520-2d20ebcf1c24.png)

**Comment by sguada** [comment]
you may need more pre_processing threads to allow building the batches for the GPUs.

**Comment by rohitgirdhar** [comment]
I am facing the exact same issue
Attached the timeline
I'm using the following code for reading data. It's slightly changed from inception example, I don't have separate threads for reading and preprocessing, but just these `shuffle_batch_join` threads that should be doing both data read and processing (`_read_image_from_disk` does the data reading and some jittering (crops/resizes))
```python
    example_list = [_read_image_from_disk(filename_queue, train) for _ in range(num_readers)]
    if train:
        example_batch, label_batch = tf.train.shuffle_batch_join(
                example_list, batch_size=batch_size,
                capacity=examples_per_shard + 3 * batch_size,
                min_after_dequeue=min_queue_examples)
    else:
        example_batch, label_batch = tf.train.batch_join(example_list, batch_size)
```


Also, tensorboard shows my queues are usually >0.7 full, which should mean the data is already there..

Would appreciate any help!

![screen shot 2016-07-25 at 10 55 04 am](https://cloud.githubusercontent.com/assets/1893429/17111425/57f9fece-5256-11e6-9fbd-244f8e9cbd71.png)

**Comment by michaelisard** [comment]
@shlens there's no Inception model owner listed: will you take a look? May be related to [tensorflow issue 3009](https://github.com/tensorflow/tensorflow/issues/3009)

**Comment by rohitgirdhar** [comment]
At least for me it seems to be related to [3009](https://github.com/tensorflow/tensorflow/issues/3009). I have added my virtual memory use details to that thread in a [comment](https://github.com/tensorflow/tensorflow/issues/3009#issuecomment-235074937)

**Comment by hetaoaoao** [comment]
After various tests, I have finally revert to tensorflow r0.9 branch, and reduce preprocess threads to 2.
Everythings works fine now, both gpu and cpu load are stable. 
On my situation, I guess my cpu&ios are still the bottleneck because gpu are not full loaded. But anyway,
current speed is fine to me.

maybe someone check out what's happening?

My host environments: i7 6700k, 2x gtx1070, 32G mem. Cuda 8rc, cudnn 5 for cuda8 rc

2016-07-30 13:52:39.932239: step 135650, loss = 5.61 (65.9 examples/sec; 0.607 sec/batch)
2016-07-30 13:52:46.062616: step 135660, loss = 5.71 (64.0 examples/sec; 0.625 sec/batch)
2016-07-30 13:52:52.152144: step 135670, loss = 5.52 (72.0 examples/sec; 0.556 sec/batch)
2016-07-30 13:52:58.461906: step 135680, loss = 4.97 (66.2 examples/sec; 0.604 sec/batch)
2016-07-30 13:53:04.639200: step 135690, loss = 4.45 (68.1 examples/sec; 0.588 sec/batch)
2016-07-30 13:53:10.830183: step 135700, loss = 4.64 (67.0 examples/sec; 0.597 sec/batch)

top - 13:59:03 up 1 day, 6 min,  3 users,  load average: 11.16, 11.35, 11.37
Tasks: 196 total,   1 running, 195 sleeping,   0 stopped,   0 zombie
%Cpu(s): 89.3 us,  2.9 sy,  0.0 ni,  7.7 id,  0.0 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem : 32826500 total,  1143460 free,  3957568 used, 27725472 buff/cache
KiB Swap: 16696316 total, 16690968 free,     5348 used. 26294200 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
20126 ateam     20   0 47.429g 5.484g 2.167g S 737.9 17.5   9182:41 python

**Comment by gunan** [comment]
Looks like the issue was resolved on master?
Closing this issue now, please reopen if you are still observing on master.
